lib/trivial.nix

```diff
@@ -355,21 +353,8 @@ rec {
      toHexString 250 => "FA"
   */
   toHexString = i:
-    let
-      toHexDigit = d:
-        if d < 10
-        then toString d
-        else
-          {
-            "10" = "A";
-            "11" = "B";
-            "12" = "C";
-            "13" = "D";
-            "14" = "E";
-            "15" = "F";
-          }.${toString d};
-    in
-      lib.concatMapStrings toHexDigit (toBaseDigits 16 i);
+    lib.concatMapStrings
+      (builtins.elemAt [ "0" "1" "2" "3" "4" "5" "6" "7" "8" "9" "A" "B" "C" "D" "E" "F" ]) (toBaseDigits 16 i);
```

Comment:
You could remove the selects (`.`) by adding these to a `let inherit` at the top. That replaces binary search by some pointer arithmetic: following pointers up the env stack and a final index to the definition.

Reply:
The list is allocated for every call, but not the elements iirc. Could move into a let binding, but after https://github.com/NixOS/nix/issues/5170 this is more efficient again. Is this function used often?

Reply:
It doesn't appear to be used often, judging by the **list** numbers in the performance report.